mlir/lib/Dialect/AIR/IR/AIRDialect.cpp

```diff
@@ -1108,6 +1109,390 @@ void WaitAllOp::getCanonicalizationPatterns(RewritePatternSet &patterns,
+      if (subviewOp != memrefOpVec.front() && !subviewOp.hasZeroOffset())
+        subviewOp->emitOpError(
+            "is not the source op in a chain of memref layout transformation "
+            "ops, but applies a non-zero offset. This feature is NYI, and "
+            "leads to unexpected behaviour.");
+    }
+  }
+
+  // Memref type at sink memref op.
+  input_ty =
+      llvm::cast<MemRefType>(memrefOpVec.back()->getResultTypes().front());
+
+  // Compose sizes and strides from the output memref type's layout.
+  strides = extractStridesFromMemrefType(input_ty, rewriter);
+  sizes = extractSizesFromMemrefType(input_ty, rewriter);
+
+  return canonicalizeAIRDmaOperands(rewriter, offsets, sizes, strides,
+                                    input_ty);
+}
+
+//
+// Dma op
+//
+
+static LogicalResult
+ComposeMemrefOpOnDmaMemcpyNdSrc(DmaMemcpyNdOp op, PatternRewriter &rewriter) {
+
+  Value memref = op.getSrcMemref();
+  if (!memref)
+    return failure();
+  auto loc = op->getLoc();
+  Value input_memref;
+  SmallVector<Value> offsets, sizes, strides;
+  offsets = op.getSrcOffsets();
+  if (!offsets.empty())
+    return failure();
+  sizes = op.getSrcSizes();
+  if (!sizes.empty())
+    return failure();
+  strides = op.getSrcStrides();
+  if (!strides.empty())
+    return failure();
+
+  if (failed(ComposeMemrefOp(memref, rewriter, input_memref, offsets, sizes,
+                             strides))) {
+    return failure();
+  }
+
+  auto newOp = rewriter.create<air::DmaMemcpyNdOp>(
+      loc, op->getResultTypes(), op.getAsyncDependencies(), op.getDstMemref(),
+      op.getDstOffsets(), op.getDstSizes(), op.getDstStrides(), input_memref,
+      offsets, sizes, strides);
+
+  for (unsigned i = 0; i < op->getNumResults(); i++) {
+    op->getResult(i).replaceAllUsesWith(newOp->getResult(i));
+  }
+
+  rewriter.eraseOp(op);
+
+  return success();
+}
+
+static LogicalResult
+ComposeMemrefOpOnDmaMemcpyNdDst(DmaMemcpyNdOp op, PatternRewriter &rewriter) {
+
+  Value memref = op.getDstMemref();
+  if (!memref)
+    return failure();
+  auto loc = op->getLoc();
+  Value input_memref;
+  SmallVector<Value> offsets, sizes, strides;
+  offsets = op.getDstOffsets();
+  if (!offsets.empty())
+    return failure();
+  sizes = op.getDstSizes();
+  if (!sizes.empty())
+    return failure();
+  strides = op.getDstStrides();
+  if (!strides.empty())
+    return failure();
+
+  if (failed(ComposeMemrefOp(memref, rewriter, input_memref, offsets, sizes,
+                             strides))) {
+    return failure();
+  }
+
+  auto newOp = rewriter.create<air::DmaMemcpyNdOp>(
+      loc, op->getResultTypes(), op.getAsyncDependencies(), input_memref,
+      offsets, sizes, strides, op.getSrcMemref(), op.getSrcOffsets(),
+      op.getSrcSizes(), op.getSrcStrides());
+
+  for (unsigned i = 0; i < op->getNumResults(); i++) {
+    op->getResult(i).replaceAllUsesWith(newOp->getResult(i));
+  }
```

Comment:
Use the rewriter instead of direct IR modification:
```suggestion
  rewriter.replaceOp(op, newOp);
```

Reply:
Updated.